### PR TITLE
kv: preemptively refresh transaction timestamps

### PR DESCRIPTION
**Pull request by nvb**
Relates to #51294.

This commit adds logic to the txnSpanRefresher to preemptively perform refreshes before issuing requests when doing is guaranteed to be beneficial. We perform a preemptive refresh if either a) doing so would be free because we have not yet accumulated any refresh spans, or b) the batch contains a committing EndTxn request that we know will be rejected if issued.

The first case is straightforward. If the transaction has yet to perform any reads but has had its write timestamp bumped, refreshing is a trivial no-op. In this case, refreshing eagerly prevents the transaction for performing any future reads at its current read timestamp. Not doing so preemptively guarantees that we will need to perform a real refresh in the future if the transaction ever performs a read. At best, this would be wasted work. At worst, this could result in the future refresh failing. So we might as well refresh preemptively while doing so is free.

Note that this first case here does NOT obviate the need for server-side refreshes. Notably, a transaction's write timestamp might be bumped in the same batch in which it performs its first read. In such cases, a preemptive refresh would not be needed but a reactive refresh would not be a trivial no-op. These situations are common for one-phase commit transactions.

The second case is more complex. If the batch contains a committing EndTxn request that we know will need a refresh, we don't want to bother issuing it just for it to be rejected. Instead, preemptively refresh before issuing the EndTxn batch. If we view reads as acquiring a form of optimistic read locks under an optimistic concurrency control scheme (as is discussed in the comment on txnSpanRefresher) then this preemptive refresh immediately before the EndTxn is synonymous with the "validation" phase of a standard OCC transaction model. However, as an optimization compared to standard OCC, the validation phase is only performed when necessary in CockroachDB (i.e. if the transaction's writes have been pushed to higher timestamps).

This second case will play into the solution for #51294. Now that we perform these preemptive refreshes when possible, we know that it is always to right choice to split off the EndTxn from the rest of the batch during a txnSpanRefresher auto-retry. Without this change, it was unclear whether the first refresh of an EndTxn batch was caused by earlier requests in the transaction or by other requests in the current batch. Now, it is always caused by other requests in the same batch, so it is always clear that we should split the EndTxn from the rest of the batch immediately after the first refresh.

Release notes (performance improvement): validation of optimistic reads is now performed earlier in transactions when doing so can save work. This eliminates certain types of transaction retry errors.

### CODE REVIEW

**Review by andreimatei — COMMENTED** [review]
Good stuff

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-:-MEyR9H80S17N0KGKF59:b-oq3t56)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyOMqR1wz2gr40zJ-B:-MEyOMqR1wz2gr40zJ-C:b-3lpmjn) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L168)):*
> ```Go
> 	ba.CanForwardReadTimestamp = sr.canForwardReadTimestampWithoutRefresh(ba.Txn)
> 
> 	// If we know that the transaction will need a refresh at some point because
> ```

As long as we're writing a long comment here (which I hope can be hidden inside `maybeRefreshPreemptively()`; see the following comment), mind also putting a TODO(andrei) explaining that whether or not we can still auto-retry at the SQL level should also play a role in deciding whether we want to refresh eagerly or not.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 395 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyKjUs-Ijrb0iUkkzN:-MEyKjUs-Ijrb0iUkkzO:bbv4pok) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L395)):*
> ```Go
> 		retryErr = preemptiveWriteTooOldErr
> 	}
> 	pErr := roachpb.NewErrorWithTxn(retryErr, ba.Txn)
> ```

I find it weird that we're creating this error here before we know that we need it (although I appreciate that we don't allocate). I'm also bothered by the fact that only a small part of `CanTransactionRetryAtRefreshedTimestamp()` seems to apply; that function doesn't seem to fit well here. For one, the no-op handling of the synthetic `WriteTooOldError` makes the timestamp flow hard to follow I think. I don't think this function should care about "write too old" at all.

I also find the splitting of responsibilities between `maybeRefreshPreemptively` and its caller pretty surprising: the caller checks that we'd want to refresh, and then this guy checks whether we can refresh. 
I would try to do a Family Guy window blinds pulling a bit to see if it'd look better with this function taking on the checking of whether we want to refresh, and on the other hand I'd leave the error creation to the caller. This method could return  `OK` or `CommitHopelessBecauseRefreshIsNeededButImpossible`. Or I dunno, I'm less sure about where the error creation should go, but one way or another I do think it should go after `tryUpdatingTxnSpans` returns false.
And then we don't have to argue about whether this function should be called `maybe` or `try` or `maybeTry`.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 426 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyR3Wb2SLVGNqy40xb:-MEyR3Wc4TviQOZSeYVr:b-wo3n7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L426)):*
> ```Go
> 	defer func() {
> 		if ok {
> 			sr.refreshSuccess.Inc(1)
> ```

I don't think we should increment this when there was nothing to refresh
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go, line 379 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyQ7XH4vWmzOrRtYVu:-MEyQ7XH4vWmzOrRtYVv:b6wae6k) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go#L379)):*
> ```Go
> // it observes a batch containing an EndTxn request that will necessarily throw
> // a serializable error.
> func TestTxnSpanrefresherPreemptiveRefresh(t *testing.T) {
> ```

s/Spanrefresher/SpanRefresher
___
*[pkg/kv/kvclient/kvcoord/txn_metrics.go, line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyQljRBK8PXOXNWpvm:-MEyQljRBK8PXOXNWpvn:b-6f2nn7) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_metrics.go#L76)):*
> ```Go
> 	metaRefreshSuccess = metric.Metadata{
> 		Name:        "txn.refresh.success",
> 		Help:        "Number of successful refreshes",
> ```

pls add to the help to link to `txn.refresh.auto_retries` and explain that some refreshes are preemptive and don't lead to a retry.
___
*[pkg/kv/kvclient/kvcoord/txn_metrics.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyHLaa1qZNw6jdBZsJ:-MEyHLab4KCLV-1iPte3:b-g7ablo) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_metrics.go#L105)):*
> ```Go
> 	metaRefreshAutoRetries = metric.Metadata{
> 		Name:        "txn.refresh.auto_retries",
> 		Help:        "Number of batch retries after successful refreshes",
> ```

nit: s/batch/request

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52884)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 168 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyOMqR1wz2gr40zJ-B:-MF29yg32sFl1AaAjcug:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L168)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

As long as we're writing a long comment here (which I hope can be hidden inside `maybeRefreshPreemptively()`; see the following comment), mind also putting a TODO(andrei) explaining that whether or not we can still auto-retry at the SQL level should also play a role in deciding whether we want to refresh eagerly or not.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 395 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyKjUs-Ijrb0iUkkzN:-MF2AAjtDB0hNOZ57y-N:b-5e7nc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L395)):*
> For one, the no-op handling of the synthetic WriteTooOldError makes the timestamp flow hard to follow I think. I don't think this function should care about "write too old" at all.

I don't quite understand what you mean here. This caller is guaranteed to hit the `TransactionRetryError` case, not the `WriteTooOldError ` case.

> I also find the splitting of responsibilities between maybeRefreshPreemptively and its caller pretty surprising: the caller checks that we'd want to refresh, and then this guy checks whether we can refresh.
> I would try to do a Family Guy window blinds pulling a bit to see if it'd look better with this function taking on the checking of whether we want to refresh, and on the other hand I'd leave the error creation to the caller. This method could return  OK or CommitHopelessBecauseRefreshIsNeededButImpossible. Or I dunno, I'm less sure about where the error creation should go, but one way or another I do think it should go after tryUpdatingTxnSpans returns false.
> And then we don't have to argue about whether this function should be called maybe or try or maybeTry.

Done.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 426 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyR3Wb2SLVGNqy40xb:-MF2AXTmBYoJXXbUJUG9:b-x5u4g5) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L426)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't think we should increment this when there was nothing to refresh
</blockquote></details>

We were before. You think we should change behavior? I tend to disagree because this still fits the definition of a "refresh". If we want to see how many refresh requests were sent, we have a metric for that too.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go, line 379 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyQ7XH4vWmzOrRtYVu:-MF2ArL81VnuC0oXAvbJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher_test.go#L379)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

s/Spanrefresher/SpanRefresher
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/txn_metrics.go, line 76 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyQljRBK8PXOXNWpvm:-MF2Bsy03_TtzYWsZLZv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_metrics.go#L76)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

pls add to the help to link to `txn.refresh.auto_retries` and explain that some refreshes are preemptive and don't lead to a retry.
</blockquote></details>

Done.
___
*[pkg/kv/kvclient/kvcoord/txn_metrics.go, line 105 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyHLaa1qZNw6jdBZsJ:-MF2B5Yq891_QUjUh-gf:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_metrics.go#L105)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: s/batch/request
</blockquote></details>

Done.

**Review by andreimatei — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52884)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei and @nvanbenschoten)

___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 395 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyKjUs-Ijrb0iUkkzN:-MF6GpWV8KDHOY0Dtmjn:b-off2ed) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L395)):*
> > For one, the no-op handling of the synthetic WriteTooOldError makes the timestamp flow hard to follow I think. I don't think this function should care about "write too old" at all.
> 
> I don't quite understand what you mean here. This caller is guaranteed to hit the TransactionRetryError case, not the WriteTooOldError  case.

Then I'm even more bothered by calling that fat function. But now that I look more, I think the `preemptiveWriteTooOldErr` can't be triggered because this interceptor terminates the `WriteTooOld` flag. Let's just assert that `ba.Txn.WriteTooOld` is not set, like we do in `sendLockedWithRefreshAttempts`.

Even if we get rid of `preemptiveWriteTooOldErr`, I'd then stil prefer to get rid of `preemptiveSerializableErr` too; I would break `CanTransactionRetryAtRefreshedTimestamp()` into two: an inner function that takes a retriable error and gives you back the timestamp that the error wants the client to retry at, and an outer function that takes an error and a timestamp and prepares the transaction for retrying at that given timestamp (i.e.
```
func prepareForRetry(txn, timestamp) (bool, txn)
if txn.CommitTimestampFixed {
  return false
}
newTxn := txn.Clone()
newTxn.WriteTimestamp.Forward(timestamp)
newTxn.ReadTimestamp.Forward(newTxn.WriteTimestamp)
newTxn.WriteTooOld = false
```

`maybeRefreshPreemptively()` would then do
`txn, ok := prepareForRetry(txn, ba.Txn.WriteTime)`

The handling of write timestamp is just too obfuscated through the `pErr := roachpb.NewErrorWithTxn(retryErr, ba.Txn)` line, for no good reason IMO.
___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 426 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyR3Wb2SLVGNqy40xb:-MF6FcTB9h0OF8QSQ0ho:b32wtd) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L426)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

We were before. You think we should change behavior? I tend to disagree because this still fits the definition of a "refresh". If we want to see how many refresh requests were sent, we have a metric for that too.
</blockquote></details>

ok

**Review by nvb — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52884)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

___
*[pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go, line 395 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-MEyKjUs-Ijrb0iUkkzN:-MF8pi171YkCggiWJ2Rt:b9u2ar0) ([raw file](https://github.com/cockroachdb/cockroach/blob/924c4ba3534da43f182af945bac95d0d4d12f961/pkg/kv/kvclient/kvcoord/txn_interceptor_span_refresher.go#L395)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

> > For one, the no-op handling of the synthetic WriteTooOldError makes the timestamp flow hard to follow I think. I don't think this function should care about "write too old" at all.
> 
> I don't quite understand what you mean here. This caller is guaranteed to hit the TransactionRetryError case, not the WriteTooOldError  case.

Then I'm even more bothered by calling that fat function. But now that I look more, I think the `preemptiveWriteTooOldErr` can't be triggered because this interceptor terminates the `WriteTooOld` flag. Let's just assert that `ba.Txn.WriteTooOld` is not set, like we do in `sendLockedWithRefreshAttempts`.

Even if we get rid of `preemptiveWriteTooOldErr`, I'd then stil prefer to get rid of `preemptiveSerializableErr` too; I would break `CanTransactionRetryAtRefreshedTimestamp()` into two: an inner function that takes a retriable error and gives you back the timestamp that the error wants the client to retry at, and an outer function that takes an error and a timestamp and prepares the transaction for retrying at that given timestamp (i.e.
```
func prepareForRetry(txn, timestamp) (bool, txn)
if txn.CommitTimestampFixed {
  return false
}
newTxn := txn.Clone()
newTxn.WriteTimestamp.Forward(timestamp)
newTxn.ReadTimestamp.Forward(newTxn.WriteTimestamp)
newTxn.WriteTooOld = false
```

`maybeRefreshPreemptively()` would then do
`txn, ok := prepareForRetry(txn, ba.Txn.WriteTime)`

The handling of write timestamp is just too obfuscated through the `pErr := roachpb.NewErrorWithTxn(retryErr, ba.Txn)` line, for no good reason IMO.

</blockquote></details>

Ok, I did pretty much exactly what you're suggesting here. I think you'll like the result. PTAL.

And if so, do you mind giving the third commit in #52885 a look while you have this paged in? That needs to land before the stability period.

**Review by andreimatei — COMMENTED** [review]
LGTM thanks

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/52884#-:-MFCGso-8YD5ZwmSyP6p:b-1cua5s)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andreimatei)

**Comment by nvanbenschoten** [comment]
api.pb.go keeps changing on me!

bors r+
